File: compatibility-tests/sdk-test-go/tests/kinesis_test.go
package tests

import (
	"context"
	"testing"
	"time"

	"floci-sdk-test-go/internal/testutil"

	"github.com/aws/aws-sdk-go-v2/aws"
	"github.com/aws/aws-sdk-go-v2/service/kinesis"
	kinesistypes "github.com/aws/aws-sdk-go-v2/service/kinesis/types"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

func TestKinesis(t *testing.T) {
	ctx := context.Background()
	svc := testutil.KinesisClient()
	streamName := "go-test-stream"

	t.Cleanup(func() {
		svc.DeleteStream(ctx, &kinesis.DeleteStreamInput{StreamName: aws.String(streamName)})
	})

	t.Run("CreateStream", func(t *testing.T) {
		_, err := svc.CreateStream(ctx, &kinesis.CreateStreamInput{
			StreamName: aws.String(streamName),
			ShardCount: aws.Int32(1),
		})
		require.NoError(t, err)
	})

	t.Run("ListStreams", func(t *testing.T) {
		r, err := svc.ListStreams(ctx, &kinesis.ListStreamsInput{})
		require.NoError(t, err)
		assert.NotEmpty(t, r.StreamNames)
	})

	var shardID string
	var streamARN string

	t.Run("DescribeStream", func(t *testing.T) {
		r, err := svc.DescribeStream(ctx, &kinesis.DescribeStreamInput{StreamName: aws.String(streamName)})
		require.NoError(t, err)
		assert.Equal(t, kinesistypes.StreamStatusActive, r.StreamDescription.StreamStatus)
		if len(r.StreamDescription.Shards) > 0 {
			shardID = aws.ToString(r.StreamDescription.Shards[0].ShardId)
		}
		streamARN = aws.ToString(r.StreamDescription.StreamARN)
	})

	t.Run("PutRecord", func(t *testing.T) {
		r, err := svc.PutRecord(ctx, &kinesis.PutRecordInput{
			StreamName:   aws.String(streamName),
			Data:         []byte(`{"event":"go-test"}`),
			PartitionKey: aws.String("pk1"),
		})
		require.NoError(t, err)
		assert.NotEmpty(t, aws.ToString(r.ShardId))
	})

	t.Run("GetShardIterator", func(t *testing.T) {
		if shardID == "" {
			t.Skip("No shard ID from DescribeStream")
		}
		r, err := svc.GetShardIterator(ctx, &kinesis.GetShardIteratorInput{
			StreamName:        aws.String(streamName),
			ShardId:           aws.String(shardID),
			ShardIteratorType: kinesistypes.ShardIteratorTypeTrimHorizon,
		})
		require.NoError(t, err)
		assert.NotEmpty(t, aws.ToString(r.ShardIterator))
	})

	t.Run("GetRecords", func(t *testing.T) {
		if shardID == "" {
			t.Skip("No shard ID from DescribeStream")
		}
		// Get fresh iterator
		iter, err := svc.GetShardIterator(ctx, &kinesis.GetShardIteratorInput{
			StreamName:        aws.String(streamName),
			ShardId:           aws.String(shardID),
			ShardIteratorType: kinesistypes.ShardIteratorTypeTrimHorizon,
		})
		require.NoError(t, err)

		r, err := svc.GetRecords(ctx, &kinesis.GetRecordsInput{
			ShardIterator: iter.ShardIterator,
			Limit:         aws.Int32(10),
		})
		require.NoError(t, err)
		assert.NotEmpty(t, r.Records)
	})

	var consumerARN string

	t.Run("RegisterStreamConsumer", func(t *testing.T) {
		if streamARN == "" {
			t.Skip("No stream ARN from DescribeStream")
		}
		r, err := svc.RegisterStreamConsumer(ctx, &kinesis.RegisterStreamConsumerInput{
			StreamARN:    aws.String(streamARN),
			ConsumerName: aws.String("go-test-consumer"),
		})
		require.NoError(t, err)
		require.NotNil(t, r.Consumer)
		consumerARN = aws.ToString(r.Consumer.ConsumerARN)
		assert.NotEmpty(t, consumerARN)
	})

	t.Run("DescribeStreamConsumer", func(t *testing.T) {
		if consumerARN == "" {
			t.Skip("No consumer ARN from RegisterStreamConsumer")
		}
		r, err := svc.DescribeStreamConsumer(ctx, &kinesis.DescribeStreamConsumerInput{
			ConsumerARN: aws.String(consumerARN),
		})
		require.NoError(t, err)
		assert.Equal(t, kinesistypes.ConsumerStatusActive, r.ConsumerDescription.ConsumerStatus)
		assert.Equal(t, "go-test-consumer", aws.ToString(r.ConsumerDescription.ConsumerName))
	})

	t.Run("SubscribeToShard", func(t *testing.T) {
		if consumerARN == "" || shardID == "" {
			t.Skip("No consumerARN or shardID")
		}
		subCtx, cancel := context.WithTimeout(ctx, 15*time.Second)
		defer cancel()

		output, err := svc.SubscribeToShard(subCtx, &kinesis.SubscribeToShardInput{
			ConsumerARN: aws.String(consumerARN),
			ShardId:     aws.String(shardID),
			StartingPosition: &kinesistypes.StartingPosition{
				Type: kinesistypes.ShardIteratorTypeTrimHorizon,
			},
		})
		require.NoError(t, err)

		stream := output.GetStream()
		defer stream.Close()

		var gotEvent bool
		for event := range stream.Events() {
			if ev, ok := event.(*kinesistypes.SubscribeToShardEventStreamMemberSubscribeToShardEvent); ok {
				gotEvent = true
				assert.NotNil(t, ev.Value.Records)
			}
		}
		require.NoError(t, stream.Err())
		assert.True(t, gotEvent, "expected at least one SubscribeToShardEvent")
	})

	t.Run("DeleteStream", func(t *testing.T) {
		_, err := svc.DeleteStream(ctx, &kinesis.DeleteStreamInput{StreamName: aws.String(streamName)})
		require.NoError(t, err)
	})
}

File: compatibility-tests/sdk-test-rust/tests/kinesis_test.rs
mod common;

use aws_sdk_kinesis::types::ShardIteratorType;

#[tokio::test(flavor = "multi_thread")]
async fn test_kinesis_create_stream() {
    let kinesis = common::kinesis_client().await;
    let stream_name = "rust-test-create-stream";

    let _guard = common::CleanupGuard::new({
        let kinesis = kinesis.clone();
        async move {
            let _ = kinesis.delete_stream().stream_name(stream_name).send().await;
        }
    });

    let result = kinesis
        .create_stream()
        .stream_name(stream_name)
        .shard_count(1)
        .send()
        .await;
    assert!(result.is_ok(), "CreateStream failed: {:?}", result.err());
}

#[tokio::test(flavor = "multi_thread")]
async fn test_kinesis_list_streams() {
    let kinesis = common::kinesis_client().await;
    let stream_name = "rust-test-list-streams";

    let _guard = common::CleanupGuard::new({
        let kinesis = kinesis.clone();
        async move {
            let _ = kinesis.delete_stream().stream_name(stream_name).send().await;
        }
    });

    // Setup
    kinesis
        .create_stream()
        .stream_name(stream_name)
        .shard_count(1)
        .send()
        .await
        .expect("setup");

    let result = kinesis.list_streams().send().await;
    assert!(result.is_ok(), "ListStreams failed: {:?}", result.err());
    assert!(!result.unwrap().stream_names().is_empty());
}

#[tokio::test(flavor = "multi_thread")]
async fn test_kinesis_describe_stream() {
    let kinesis = common::kinesis_client().await;
    let stream_name = "rust-test-describe-stream";

    let _guard = common::CleanupGuard::new({
        let kinesis = kinesis.clone();
        async move {
            let _ = kinesis.delete_stream().stream_name(stream_name).send().await;
        }
    });

    // Setup
    kinesis
        .create_stream()
        .stream_name(stream_name)
        .shard_count(1)
        .send()
        .await
        .expect("setup");

    let result = kinesis.describe_stream().stream_name(stream_name).send().await;
    assert!(result.is_ok(), "DescribeStream failed: {:?}", result.err());
}

#[tokio::test(flavor = "multi_thread")]
async fn test_kinesis_put_record() {
    let kinesis = common::kinesis_client().await;
    let stream_name = "rust-test-put-record";

    let _guard = common::CleanupGuard::new({
        let kinesis = kinesis.clone();
        async move {
            let _ = kinesis.delete_stream().stream_name(stream_name).send().await;
        }
    });

    // Setup
    kinesis
        .create_stream()
        .stream_name(stream_name)
        .shard_count(1)
        .send()
        .await
        .expect("setup");

    let result = kinesis
        .put_record()
        .stream_name(stream_name)
        .data(aws_smithy_types::Blob::new(b"{\"event\":\"rust-test\"}".to_vec()))
        .partition_key("pk1")
        .send()
        .await;
    assert!(result.is_ok(), "PutRecord failed: {:?}", result.err());
    assert!(!result.unwrap().shard_id().is_empty());
}

#[tokio::test(flavor = "multi_thread")]
async fn test_kinesis_get_records() {
    let kinesis = common::kinesis_client().await;
    let stream_name = "rust-test-get-records";

    let _guard = common::CleanupGuard::new({
        let kinesis = kinesis.clone();
        async move {
            let _ = kinesis.delete_stream().stream_name(stream_name).send().await;
        }
    });

    // Setup
    kinesis
        .create_stream()
        .stream_name(stream_name)
        .shard_count(1)
        .send()
        .await
        .expect("setup");

    // Put a record
    kinesis
        .put_record()
        .stream_name(stream_name)
        .data(aws_smithy_types::Blob::new(b"test data".to_vec()))
        .partition_key("pk1")
        .send()
        .await
        .expect("put record");

    // Get shard ID
    let desc = kinesis
        .describe_stream()
        .stream_name(stream_name)
        .send()
        .await
        .expect("describe");
    let shard_id = desc
        .stream_description()
        .and_then(|d| d.shards().first())
        .map(|s| s.shard_id())
        .unwrap_or("");

    // Get shard iterator
    let iter = kinesis
        .get_shard_iterator()
        .stream_name(stream_name)
        .shard_id(shard_id)
        .shard_iterator_type(ShardIteratorType::TrimHorizon)
        .send()
        .await
        .expect("get iterator");

    let shard_iterator = iter.shard_iterator().unwrap_or("");

    // Get records
    let result = kinesis
        .get_records()
        .shard_iterator(shard_iterator)
        .limit(10)
        .send()
        .await;
    assert!(result.is_ok(), "GetRecords failed: {:?}", result.err());
    assert!(!result.unwrap().records().is_empty());
}

#[tokio::test(flavor = "multi_thread")]
async fn test_kinesis_delete_stream() {
    let kinesis = common::kinesis_client().await;
    let stream_name = "rust-test-delete-stream";

    // No cleanup guard needed - test is about deletion

    // Setup
    kinesis
        .create_stream()
        .stream_name(stream_name)
        .shard_count(1)
        .send()
        .await
        .expect("setup");

    let result = kinesis.delete_stream().stream_name(stream_name).send().await;
    assert!(result.is_ok(), "DeleteStream failed: {:?}", result.err());
}
